sys/linux/devio.txt

```diff
@@ -0,0 +1,129 @@
+ioctl$USBDEVFS_SETINTERFACE(fd fd_devio, cmd const[USBDEVFS_SETINTERFACE], arg ptr[in, usbdevfs_setinterface])
+ioctl$USBDEVFS_SETCONFIGURATION(fd fd_devio, cmd const[USBDEVFS_SETCONFIGURATION], arg ptr[in, int32])
+ioctl$USBDEVFS_SUBMITURB(fd fd_devio, cmd const[USBDEVFS_SUBMITURB], arg ptr[in, usbdevfs_urb])
+ioctl$USBDEVFS_DISCARDURB(fd fd_devio, cmd const[USBDEVFS_DISCARDURB], arg ptr[in, int64])
+ioctl$USBDEVFS_DISCSIGNAL(fd fd_devio, cmd const[USBDEVFS_DISCSIGNAL], arg ptr[in, usbdevfs_disconnectsignal])
+ioctl$USBDEVFS_CLAIMINTERFACE(fd fd_devio, cmd const[USBDEVFS_CLAIMINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASEINTERFACE(fd fd_devio, cmd const[USBDEVFS_RELEASEINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_IOCTL(fd fd_devio, cmd const[USBDEVFS_IOCTL], arg ptr[in, usbdevfs_ioctl])
+ioctl$USBDEVFS_CLAIM_PORT(fd fd_devio, cmd const[USBDEVFS_CLAIM_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASE_PORT(fd fd_devio, cmd const[USBDEVFS_RELEASE_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_CAPABILITIES(fd fd_devio, cmd const[USBDEVFS_GET_CAPABILITIES], arg ptr[out, int32])
+ioctl$USBDEVFS_DISCONNECT_CLAIM(fd fd_devio, cmd const[USBDEVFS_DISCONNECT_CLAIM], arg ptr[in, usbdevfs_disconnect_claim])
+ioctl$USBDEVFS_FREE_STREAMS(fd fd_devio, cmd const[USBDEVFS_FREE_STREAMS], arg ptr[in, usbdevfs_streams])
+ioctl$USBDEVFS_DROP_PRIVILEGES(fd fd_devio, cmd const[USBDEVFS_DROP_PRIVILEGES], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_SPEED(fd fd_devio, cmd const[USBDEVFS_GET_SPEED])
+
+usbdevfs_ctrltranfer {
+	bRequestType	int8
+	bRequest	int8
+	wValue		int16
+	wIndex		int16
+	wLength		int16
+	timeout		int32
+	data		ptr[inout, array[int8]]
+}
+
+usbdevfs_bulktransfer {
+	ep	int32
+	len	len[data, int32]
+	timeout	int32
+	data	ptr[inout, array[int8]]
+}
+
+usbdevfs_getdriver {
+	interface	int32
+	driver		ptr[out, array[int8, MAXDRIVERNAME]]
+}
+
+usbdevfs_connectinfo {
+	devnum	int32
+	slow	int8
+}
+
+usbdevfs_setinterface {
+	interface	int32
+	altsetting	int32
+}
+
+usbdevfs_urb {
+	type		flags[urb_types, int8]
+	endpoint	int8
+	status		int32
+	flags		flags[urb_flags, int32]
+	buffer		ptr[in, array[int8]]
+	buffer_length	len[buffer, int32]
+	actual_length	int32
+	start_frame	int32
+	u		usbdevfs_urb_union
+	error_count	int32
+	signr		int32
+	usercontext	ptr[in, array[int8]]
+	iso_frame_desc	array[usbdevfs_iso_packet_desc]
+}
+
+urb_types = USBDEVFS_URB_TYPE_ISO, USBDEVFS_URB_TYPE_INTERRUPT, USBDEVFS_URB_TYPE_CONTROL, USBDEVFS_URB_TYPE_BULK
+urb_flags = USBDEVFS_URB_SHORT_NOT_OK, USBDEVFS_URB_ISO_ASAP, USBDEVFS_URB_BULK_CONTINUATION, USBDEVFS_URB_NO_FSBR, USBDEVFS_URB_ZERO_PACKET, USBDEVFS_URB_NO_INTERRUPT
+
+usbdevfs_urb_union [
+	number_of_packets	int32
+	stream_id		int32
+]
+
+usbdevfs_iso_packet_desc {
+	length		int32
+	actual_length	int32
+	status		int32
+}
+
+usbdevfs_disconnectsignal {
+	signr	int32
+	context	ptr[in, array[int8]]
+}
+
+usbdevfs_ioctl {
```

Comment:
Let's make this a union of 3 structs, `usbdevfs_ioctl_disconnect` for `USBDEVFS_DISCONNECT`, `usbdevfs_ioctl_connect` for `USBDEVFS_CONNECT` (with precisely defined `ioctl_code` and `data`), and `usbdevfs_ioctl_driver` for generic driver ioctls.

https://elixir.bootlin.com/linux/v5.3.1/source/drivers/usb/core/devio.c#L2151